### PR TITLE
remove .eval() from some rvalue functions

### PR DESCRIPTION
**Pull request by SteveBronder**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary

Since we can pass around expressions now this removes `.eval()` from some `rvalue()` functions that do not need it. Sadly we still have to keep the `.eval()` for `min_max_index` types since they can return a `Block<>` or a `Reverse<Block<>>`

#### Intended Effect

Allow slices to pass through to stan math functions

#### How to Verify

Tests in rvalue

#### Side Effects

Passing Eigen expressions to the math library

#### Documentation

Docs already there

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Steve Bronder

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: src/stan/model/indexing/rvalue.hpp
Comment (bbbales2) [inline src/stan/model/indexing/rvalue.hpp:381]: There are evals in this function. Is that okay?

File: src/stan/model/indexing/rvalue.hpp
Comment (bbbales2) [inline src/stan/model/indexing/rvalue.hpp:726]: Expressions not work here? Or are they of different types so they have to be eval'd?

**Review by bbbales2 — CHANGES_REQUESTED** [review]
Code changes look good. Did we start returning expressions up here? Or is this step 1 in doing expressions in stan-dev/stan?

I know this pull exists: https://github.com/stan-dev/math/pull/2190, so I know we're intent on returning expressions once tests pass, but I didn't know if there were any dangling requirements for expression functions or anything (like things that need to have holders and whatnot).

@t4c1 for comment

File: src/stan/model/indexing/rvalue.hpp
Comment (SteveBronder) [reply src/stan/model/indexing/rvalue.hpp:381]: Yeah so for all of the `min_max` functions we need to evaluate before returning because in one case we would return a `Block<>` and in the other case we would return a block with `Reverse<Block<>>`

File: src/stan/model/indexing/rvalue.hpp
Comment (SteveBronder) [reply src/stan/model/indexing/rvalue.hpp:726]: Yep same thing here with min_max

**Comment by SteveBronder** [comment]
> I know this pull exists: stan-dev/math#2190, so I know we're intent on returning expressions once tests pass, but I didn't know if there were any dangling requirements for expression functions or anything (like things that need to have holders and whatnot).

> @t4c1 for comment

Yeah I'm actually not sure if these need holders or not. Tadej will pretty much anything returning an expression to be wrapped in holder?

**Comment by t4c1** [comment]
Well we are still waiting on @rok-cesnovar to open that PR with user defined functions supporting expressions.

Most functions do not need a holder. Only functions returning an expression that references a local variable plain matrix do.

**Review by t4c1 — APPROVED** [review]
Looks good. It just need to wait on @rok-cesnovar before it is merged. @rok-cesnovar How are expressions in UDFs comming along?

**Comment by rok-cesnovar** [comment]
Should be finally done today (famous last words).

**Comment by rok-cesnovar** [comment]
Good to merge now right?

**Review by bbbales2 — APPROVED** [review]

**Comment by bbbales2** [comment]
@rok-cesnovar if you're done with your stuff, sure thing

**Comment by rok-cesnovar** [comment]
Yep UDFs accepting expressions is live.

**Pull request merged**
